### PR TITLE
Add azure_policy_definition and azure_policy_set_definition entities

### PR DESCRIPTION
**Pull request by ndowmon**
Screen Shot 2021-04-12 at 11 26 47 AM

### CODE REVIEW

File: src/steps/resource-manager/policy/client.ts [hunk 1]
```diff
@@ -27,4 +31,103 @@ export class AzurePolicyClient extends Client {
...
+        },
+        'Error calling policyDefinitions.getBuiltIn',
+      );
+    }
+  }
+}
+
+function getNameFromPolicyDefinitionId(policyDefinitionId: string) {
+  const slashDelimetedSegements = policyDefinitionId.split('/');
+  return slashDelimetedSegements[slashDelimetedSegements.length - 1];
```
Comment (mknoedel) [inline src/steps/resource-manager/policy/client.ts:132]: Unimportant comment: slightly risky if policyDefinition is ''

File: src/steps/resource-manager/policy/findOrCreatePolicyDefinitionEntityFromAssignment.ts [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+    return;
+  }
+
+  switch (type) {
+    case PolicyDefinitionType.POLICY_DEFINITION: {
+      const policyDefinition = await getPolicyDefinition(
+        client,
+        policyDefinitionId,
+      );
+      if (!policyDefinition) return; // TODO warn message
```
Comment (mknoedel) [inline src/steps/resource-manager/policy/findOrCreatePolicyDefinitionEntityFromAssignment.ts:97]: TODO

File: src/steps/resource-manager/policy/findOrCreatePolicyDefinitionEntityFromAssignment.ts [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+            policyDefinitionReferenceId:
+              policyDefinitionReference.policyDefinitionReferenceId,
+            // the following properties have invalid types for edges:
+            // parameters: policyDefinitionReference.parameters,
+            // groupNames: policyDefinitionReference.groupNames,
+          },
+        }),
+      );
+    } else {
+      // TODO warn
```
Comment (mknoedel) [inline src/steps/resource-manager/policy/findOrCreatePolicyDefinitionEntityFromAssignment.ts:179]: TODO

File: src/steps/resource-manager/policy/findOrCreatePolicyDefinitionEntityFromAssignment.ts [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+    }
+  }
+}
+
+/**
+ * Built-in definitions do not start with `/subscriptions/${subscription-id}/`, while custom definitions do.
```
Comment (mknoedel) [inline src/steps/resource-manager/policy/findOrCreatePolicyDefinitionEntityFromAssignment.ts:185]: 👍

File: src/steps/resource-manager/policy/client.ts [hunk 5]
```diff
@@ -27,4 +31,103 @@ export class AzurePolicyClient extends Client {
...
+    const serviceClient = await this.getAuthenticatedServiceClient(
+      PolicyClient,
+    );
+
+    const name = getNameFromPolicyDefinitionId(policySetDefinitionId);
+    try {
+      const response = await serviceClient.policySetDefinitions.get(name);
+      return response;
+    } catch (err) {
+      this.logger.warn(
```
Comment (mknoedel) [inline src/steps/resource-manager/policy/client.ts:47]: Should we throw an error here instead?

File: src/steps/resource-manager/policy/client.ts (hunk 1 above)
Comment (mknoedel) [reply src/steps/resource-manager/policy/client.ts:132]: or undefined

File: src/steps/resource-manager/policy/client.ts [hunk 6]
```diff
@@ -27,4 +31,103 @@ export class AzurePolicyClient extends Client {
...
+    const serviceClient = await this.getAuthenticatedServiceClient(
+      PolicyClient,
+    );
+
+    const name = getNameFromPolicyDefinitionId(policyDefinitionId);
+    try {
+      const response = await serviceClient.policyDefinitions.get(name);
+      return response;
+    } catch (err) {
+      this.logger.warn(
```
Comment (mknoedel) [inline src/steps/resource-manager/policy/client.ts:95]: Also maybe throw an error here instead?

File: src/steps/resource-manager/policy/index.test.ts [hunk 7]
```diff
@@ -19,60 +24,223 @@ afterEach(async () => {
...
+    expect(policyAssignmentEntities.length).toBeGreaterThan(0);
+    expect(policyAssignmentEntities).toMatchGraphObjectSchema({
+      _class: PolicyEntities.POLICY_ASSIGNMENT._class,
+    });
+
+    expect(context.jobState.collectedRelationships).toHaveLength(0);
   });
+});
+
+describe.only('rm-policy-definitions', () => {
```
Comment (mknoedel) [inline src/steps/resource-manager/policy/index.test.ts:64]: Need to remove "only"

File: src/steps/resource-manager/policy/index.test.ts [hunk 8]
```diff
@@ -19,60 +24,223 @@ afterEach(async () => {
...
-      _key: `/subscriptions/${instanceConfig.subscriptionId}/resourceGroups/j1dev-policy-resource-group|has|/subscriptions/${instanceConfig.subscriptionId}/resourceGroups/j1dev-policy-resource-group/providers/Microsoft.Authorization/policyAssignments/j1dev-policy-assignment`,
-      _type: 'azure_resource_has_policy_assignment',
-      _class: 'HAS',
-      _fromEntityKey: `/subscriptions/${instanceConfig.subscriptionId}/resourceGroups/j1dev-policy-resource-group`,
-      _toEntityKey: `/subscriptions/${instanceConfig.subscriptionId}/resourceGroups/j1dev-policy-resource-group/providers/Microsoft.Authorization/policyAssignments/j1dev-policy-assignment`,
-      displayName: 'HAS',
-    }),
-  );
+    expect(restRelationships).toHaveLength(0);
+  }, 150000);
```
Comment (mknoedel) [inline src/steps/resource-manager/policy/index.test.ts:245]: That's a long timer

**Review by mknoedel — COMMENTED** [review]
Great work!  a few comments

File: src/steps/resource-manager/policy/index.test.ts (hunk 8 above)
Comment (ndowmon) [reply src/steps/resource-manager/policy/index.test.ts:245]: yeah 😞  this test makes a ton of API calls...

File: src/steps/resource-manager/policy/client.ts (hunk 5 above)
Comment (ndowmon) [reply src/steps/resource-manager/policy/client.ts:47]: Yes I think that's a good idea.

**Review by mknoedel — APPROVED** [review]
Looks great!

**Pull request merged**